### PR TITLE
Burn sys_fee

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

File: src/neo/SmartContract/Native/Tokens/GasToken.cs [hunk 1]
```diff
@@ -44,30 +37,7 @@ protected override bool OnPersist(ApplicationEngine engine)
             ECPoint[] validators = NEO.GetNextBlockValidators(engine.Snapshot);
             UInt160 primary = Contract.CreateSignatureRedeemScript(validators[engine.Snapshot.PersistingBlock.ConsensusData.PrimaryIndex]).ToScriptHash();
             Mint(engine, primary, engine.Snapshot.PersistingBlock.Transactions.Sum(p => p.NetworkFee));
-            BigInteger sys_fee = GetSysFeeAmount(engine.Snapshot, engine.Snapshot.PersistingBlock.Index - 1) + engine.Snapshot.PersistingBlock.Transactions.Sum(p => p.SystemFee);
```
Comment (shargon) [inline src/neo/SmartContract/Native/Tokens/GasToken.cs]: Also we should decrement the `Total Supply` otherwise people won't know the inflation

Comment (shargon) [reply src/neo/SmartContract/Native/Tokens/GasToken.cs]: I need to re-check maybe it's done already here https://github.com/neo-project/neo/pull/1430/files#diff-721ebd5f13602811b4f33d4a51a354d7R36

**Comment by i359** [comment]
Burn addr is ?  how to view burn record? thanks!

File: src/neo/Network/P2P/Payloads/Transaction.cs [hunk 2]
```diff
@@ -127,7 +127,7 @@ public int Size
         }
 
         /// <summary>
-        /// Distributed to NEO holders.
+        /// Fee to be burned.
```
Comment (vncoelho) [inline src/neo/Network/P2P/Payloads/Transaction.cs]: Good comment, @shargon.

**Comment by superboyiii** [comment]
@erikzhang Is this ready for test?

**Comment by Tommo-L** [comment]
We have not yet decided how to distribute GAS

**Comment by erikzhang** [comment]
> We have not yet decided how to distribute GAS

We don't need to distribute sysfee anymore.

File: src/neo/SmartContract/Native/Tokens/GasToken.cs [hunk 3]
```diff
@@ -44,30 +37,7 @@ protected override bool OnPersist(ApplicationEngine engine)
...
-        [ContractMethod(0_01000000, ContractParameterType.Integer, ParameterTypes = new[] { ContractParameterType.Integer }, ParameterNames = new[] { "index" }, SafeMethod = true)]
-        private StackItem GetSysFeeAmount(ApplicationEngine engine, VMArray args)
-        {
-            uint index = (uint)args[0].GetBigInteger();
-            return GetSysFeeAmount(engine.Snapshot, index);
-        }
-
-        public BigInteger GetSysFeeAmount(StoreView snapshot, uint index)
-        {
-            if (index == 0) return Blockchain.GenesisBlock.Transactions.Sum(p => p.SystemFee);
```
Comment (Tommo-L) [inline src/neo/SmartContract/Native/Tokens/GasToken.cs]: Should we keep the `SysFeeAmount` ? If we want to increase the circulation of gas, will it depend on the `SysFeeAmount`?

**Comment by shargon** [comment]
> We don't need to distribute sysfee anymore.

Then NEO holders won't receive GAS with the stacking?

**Comment by vncoelho** [comment]
@shargon, if we burn sys_fee, HOLDERS will receive the Inflation only.

**Comment by erikzhang** [comment]
> if we burn sys_fee, HOLDERS will receive the Inflation only.

If we burn sys_fee, the total supply of GAS decreases, and the price of GAS should raise. This benefits NEO holders indirectly.

**Comment by shargon** [comment]
> If we burn sys_fee, the total supply of GAS decreases, and the price of GAS should raise. This benefits NEO holders indirectly.

This benefits GAS holders, because NEO holders doesn't receive GAS.

**Comment by erikzhang** [comment]
> This benefits GAS holders, because NEO holders doesn't receive GAS.

NEO holders receive GAS every block. (The inflation part.)

**Comment by Tommo-L** [comment]
Acceptable.  And considering that our contract price has not been reduced, I hope we can add the storage payback mechanism(rental model)

**Comment by shargon** [comment]
@erikzhang Can we join both fees now, and subtract the exact quantity that we need to burn, in order to reduce the TX fields.

**Comment by erikzhang** [comment]
> Can we join both fees now, and subtract the exact quantity that we need to burn, in order to reduce the TX fields.

I don't think this is browser friendly. Maybe we can use `int` to reduce the bytes. In this case, the fee is up to 20 GAS.

**Comment by vncoelho** [comment]
@erikzhang, I like the idea of merging the fees.

I believe Igor and Wang Yong might also be in favor, because we discussed this before.

**Comment by vang1ong7ang** [comment]
> > If we burn sys_fee, the total supply of GAS decreases, and the price of GAS should raise. This benefits NEO holders indirectly.
> 
> This benefits GAS holders, because NEO holders doesn't receive GAS.

I understand what shargon means. It is reasonable.

Actually distributing sys_fee to NEO holders may makes NEO holders benefits more. Burning sys_fee benefits GAS holders **directly**, and benefits NEO holders **indirectly**. The abstract values of benefit depend on the rate of inflation of GAS and the amount of sys_fee.

**Comment by vang1ong7ang** [comment]
I like the idea of merging the fees, too.

Using the same way to distribute sys_fee and net_fee makes the design simple and anti-spam, and I don't see enough disadvantages.

**Comment by Tommo-L** [comment]
For me, burn system fee is acceptable, and we also need to support dynamic adjustment fee when GAS price is too high.

**Comment by erikzhang** [comment]
Does anyone have an objection?

**Comment by nicolegys** [comment]
Test passed.

Test steps:
1. send neo from multisigaddr to single consensus addr, then show gas.
---when burning gas---
image
---when not burning gas---
image
---getrawtransaction---
image
We can see the sysfee is 1gas, consistent with the return result of neo-cli.

2.send neo to itself, then list asset&show gas.
---when burning gas---
image
---when not burning gas---
image
We can see the gas amout when burning is 1 less than that when not burning, the same for the amount of unclaimed gas.

**Pull request merged**